fem/fe.cpp

```diff
@@ -1377,10 +1427,58 @@ void VectorFiniteElement::LocalInterpolation_RT(
    }
 }
 
+void VectorFiniteElement::LocalL2Projection_ND(
+   const VectorFiniteElement &cfe,
+   ElementTransformation &Trans, DenseMatrix &I) const
+{
+   Vector v(dim);
+   IntegrationPoint tr_ip;
+
+   const int fs = dof, cs = cfe.GetDof();
+   I.SetSize(fs, cs);
+   DenseMatrix fine_shape(fs, dim), coarse_shape(cs, cfe.GetDim());
+   DenseMatrix fine_mass(fs), fine_coarse_mass(fs, cs); // initialized with 0
+   const int ir_order =
+      std::max(GetOrder(), this->GetOrder()) + this->GetOrder();
+   const IntegrationRule &ir = IntRules.Get(this->GetGeomType(), ir_order);
+
+   Trans.SetIntPoint(&Geometries.GetCenter(geom_type));
+   const DenseMatrix &J = Trans.Jacobian();
+   for (int i = 0; i < ir.GetNPoints(); i++)
+   {
+      const IntegrationPoint &ip = ir.IntPoint(i);
+      this->CalcVShape(ip, fine_shape);
+      Trans.Transform(ip, v);
+      tr_ip.Set(v, dim);
+      cfe.CalcVShape(tr_ip, coarse_shape);
+
+      AddMult_a_AAt(ip.weight, fine_shape, fine_mass);
+      for (int k=0; k<fs; ++k)
+      {
+         for (int j=0; j<cs; ++j)
+         {
+            double Mkj = 0.0;
+            for (int d1=0; d1<dim; ++d1)
+            {
+               for (int d2=0; d2<dim; ++d2)
+               {
+                  Mkj += ip.weight*fine_shape(k,d1)*J(d1,d2)*coarse_shape(j,d2);
+               }
+            }
+            fine_coarse_mass(k,j) += (fabs(Mkj) < 1e-12) ? 0.0 : Mkj;
+         }
+      }
+   }
+   DenseMatrixInverse fine_mass_inv(fine_mass);
+   fine_mass_inv.Mult(fine_coarse_mass, I);
+}
+
 void VectorFiniteElement::LocalInterpolation_ND(
    const VectorFiniteElement &cfe, const double *tk, const Array<int> &d2t,
    ElementTransformation &Trans, DenseMatrix &I) const
 {
+   if (!is_nodal) { return LocalL2Projection_ND(cfe, Trans, I); }
```

Comment:
In the nodal case, this function just does interpolation, so why introduce L2 projection? Alternatively, could we do something similar to the `ProjectIntegrated` functions, which do integrated interpolation? This would be specialized in classes such as `ND_QuadrilateralElement`, which would call its own function instead of this base class function. I am not saying we definitely should do this, but some questions are (i) why do L2 projection vs. interpolation; (ii) is it clear to the user whether projection or interpolation is done in each case; (iii) is L2 projection too expensive or unstable for very high order? My initial thoughts are (i) L2 projection is more accurate, but it is not clear why this choice is made here other than it is general for various derived classes; (ii) it is not clear to the user, and mixing projection and interpolation in different cases may cause confusion or problems; (iii) for moderate order (with PA we are limited to about order 5 or 6 for 3D Nedelec anyway), the dense matrix inversion should be fine, but for high enough order the inversion may get expensive and possibly inaccurate/unstable.

Reply:
Thanks, this makes sense. It seems we agree that possible future work (after this PR) may be to add special cases of integrated interpolation for some derived classes to replace the general L2 projection from the base class, and to improve the dense matrix inversion for high order.